dialogy/base/plugin/__init__.py

```diff
@@ -274,24 +271,75 @@ def __call__(self, workflow: Workflow) -> None:
         - The workflow takes care of keeping its :ref:`input<Input>` and :ref:`output<Output>` immutable.
 
         .. _PluginBookkeeping:
-
+        
         :param workflow: An instance of :ref:`Workflow <WorkflowClass>`.
         :type workflow: Workflow
         """
         logger.enable(str(self)) if self.debug else logger.disable(str(self))
 
-        if workflow.input is None:
-            return
+        if input is None:
+            return input, output
 
-        if workflow.output is None:
-            return
+        if output is None:
+            return input, output
 
-        if self.prevent(workflow.input, workflow.output):
-            return
+        if self.prevent(input, output):
+            return input, output
 
-        value = self.utility(workflow.input, workflow.output)
+        # compute
+        value = self.utility(input, output)
+
+        # update
         if value is not None and isinstance(self.dest, str):
-            workflow.set(self.dest, value, self.replace_output)
+            input, output = self.set(value, input, output)
+        
+        return input, output
+
+
+    def set(
```

Comment:
I think the only issue with doing this would be that a lot of 'setting' logic is leaking into the plugin class which doesn't seem like good design.

In a setting where plugins are run asynchronously in parallel, this would mean that there is additional logic to reconcile the output fetched from multiple plugins and that output could be clashing as well? Or are you planning on just simply aggregating the output across the plugins run in parallel by a simple merge approach?

Reply:
> 'setting' logic is leaking into the plugin class doesn't seem like good design.

I feel it's better than having it in Workflow but calling it from Plugin which then modifies internal state of Workflow. This is a lot cleaner, that the Plugin has independence and there is congruence between its input/output. Especially now that we've moved a few plugins out to another service where they don't have access to Workflow. This also makes the API request/response handling simpler.

> In a setting where plugins are run asynchronously in parallel

That would require rethinking a lot of things. I'm intentionally not taking DAG or other optimisations into consideration for now.